File: python/problem_65.py
from typing import List


def spiral_order(matrix: List[List[int]]) -> List[int]:
    if not matrix:
        return []
    top, bottom, left, right = 0, len(matrix)-1, 0, len(matrix[0])-1
    result = []
    while top <= bottom and left <= right:
        for c in range(left, right+1): result.append(matrix[top][c])
        top += 1
        for r in range(top, bottom+1): result.append(matrix[r][right])
        right -= 1
        if top <= bottom:
            for c in range(right, left-1, -1): result.append(matrix[bottom][c])
            bottom -= 1
        if left <= right:
            for r in range(bottom, top-1, -1): result.append(matrix[r][left])
            left += 1
    return result


def main():
    print("Problem 65:")
    m = [[1,2,3,4,5],[6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]]
    assert spiral_order(m) == [1,2,3,4,5,10,15,20,19,18,17,16,11,6,7,8,9,14,13,12]
    print("All Tests Passed")

File: typescript/problem_65.ts
export function spiralOrder(matrix: number[][]): number[] {
    let top=0, bottom=matrix.length-1, left=0, right=matrix[0].length-1;
    const result: number[] = [];
    while (top<=bottom && left<=right) {
        for (let c=left;c<=right;c++) result.push(matrix[top][c]); top++;
        for (let r=top;r<=bottom;r++) result.push(matrix[r][right]); right--;
        if (top<=bottom) { for (let c=right;c>=left;c--) result.push(matrix[bottom][c]); bottom--; }
        if (left<=right) { for (let r=bottom;r>=top;r--) result.push(matrix[r][left]); left++; }
    }
    return result;
}

export function problem65(): void {
    console.log("Problem 65:");
    const m = [[1,2,3,4,5],[6,7,8,9,10],[11,12,13,14,15],[16,17,18,19,20]];
    const exp = [1,2,3,4,5,10,15,20,19,18,17,16,11,6,7,8,9,14,13,12];
    if (JSON.stringify(spiralOrder(m)) !== JSON.stringify(exp)) throw new Error("assertion failed");
    console.log("All Tests Passed");
}
